### DISCUSSION TITLE
React-like context for the server

### DISCUSSION THREAD

**Discussion by kvnwolf** [vercel/next.js#51808] [Ideas]
### Goals

1. Allow to wrap a set of routes with a context
2. Reduce code duplication (i.e. I wrap my authenticated routes with a `currentUser` context and I don't need to make further API calls)
3. Ideally, every code that is processed by the server (actions / RSC) can consume the context

### Non-Goals

1. To use the context in client components

### Background

Currently, I am working on an application where I am seeing the need to fetch the same data multiple times across the same tree (i.e. current user data). I think that being able to create a context, and then just wrap my child routes could be a solution to avoid repeating the same query multiple times.

I was inspecting the Next.js code and found that it [is already using `AsyncLocalStorage`](https://github.com/vercel/next.js/blob/70aad4a2b5b34fbbae006c477f96c54118da6381/packages/next/src/server/app-render/app-render.tsx#L1669), when calling functions like `cookies()`, `headers()`, etc. So I _think_ implementation maybe won't be too difficult.

### Proposal

It could be another special file of the app folder, at the same level of `layout.tsx`, `page.tsx` or `route.ts`. Maybe something like `context.ts`, which will wrap all the child routes (including the layout).

**Comment by oliviertassinari** [comment]
Could it be a duplicate of #42301?
